server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java

```diff
@@ -97,68 +138,88 @@ final class BulkOperation extends ActionRunnable<BulkResponse> {
         this.relativeTimeProvider = relativeTimeProvider;
         this.indexNameExpressionResolver = indexNameExpressionResolver;
         this.client = client;
-        this.observer = new ClusterStateObserver(clusterService, bulkRequest.timeout(), logger, threadPool.getThreadContext());
+        this.observer = observer;
     }
 
     @Override
     protected void doRun() {
         assert bulkRequest != null;
         final ClusterState clusterState = observer.setAndGetObservedState();
-        if (handleBlockExceptions(clusterState)) {
+        if (handleBlockExceptions(clusterState, BulkOperation.this, this::onFailure)) {
+            return;
+        }
+        Map<ShardId, List<BulkItemRequest>> requestsByShard = groupBulkRequestsByShards(clusterState);
+        executeBulkRequestsByShard(requestsByShard, clusterState, this::redirectFailuresOrCompleteBulkOperation);
+    }
+
+    private void doRedirectFailures() {
+        assert failureStoreRedirects.isEmpty() != true : "Attempting to redirect failures, but none were present in the queue";
+        final ClusterState clusterState = observer.setAndGetObservedState();
+        if (handleBlockExceptions(clusterState, ActionRunnable.run(listener, this::doRedirectFailures), this::discardRedirectsAndFinish)) {
```

Comment:
I assume we discard redirects because we don't want to start doing more stuff (i.e. the failure store writes) when the cluster is blocked? Is that maybe worth a comment or do you think that's clear enough already for non-rookies?

Reply:
I have added a javadoc on `handleBlockException` and a comment on this line to make it more clear